2023-07-28-on-demand-builders.md

```diff
@@ -0,0 +1,125 @@
+On demand started builders in Copr
+==================================
+
+The problem is that our builders are too slow to build e.g. Chromium package
+quickly enough.  Such builds take 24h and more (our limit is 48h).  With the
+cloud era, it really makes sense to have more powerful machines -> even though
+they cost more $$/hour, we spend less amount of time on them.
+
+The problem
+-----------
+
+The thing is that, as of now, every builder in Copr is started by Resalloc "in
+advance", or "preallocated".  The builders are preallocated even though they are
+not _immediately_ used, and stay idle till someone takes them (if there's no
+demand for days, they stay allocated).  When taken, system allocates additional
+resources (so we get more and more machines if there's actual demand for them).
+
+This system works well for normal builders, because they are typically taken
+very quickly (the more tasks in queue the faster they are taken; the smaller the
+"prealloc" quota number for resalloc pool is the faster they are taken).  And
+even if they are not taken for a while, we can actually afford keeping a few of
+them allocated because the instances are not overly expensive.
+
+There's though a demand to have a more (maybe ridiculously) powerful instances;
+having just one of them allocated but unused could be quite expensive matter.
+Even if we used for a low-demanding task.
+
+
+The proposal overview
+---------------------
+
+Let's have on-demand resources/pools in Resalloc.  In such pools there would be
+**no preallocated resource** by default, as long as there's no demand for them.
+Client will have a possibility to ask for them via special "on demand" tagged
+tickets.
+
+We want to be able to
+
+    - allow-list for supported architectures (because for some
+      architectures like s390x it might be quite hard to get "extra_powerful"
+      machines).
+
+    - allowlist the sets of users that can use the tags, preferably
+      through OIDC/FAS/LDAP groups -> we might want to have some acknowledge
+      from those users that the feature won't be overused or misused
```

Comment:
We will probably want to also limit max number of beefy builders per one user/project/whatever.

Reply:
This is exactly what I want to avoid. And not trying to solve at all.

Reply:
Well, isn't this just a misunderstanding? This already seems like rather a configuration issue only.  We have some limits implemented (solving similar problems) I believe: https://github.com/fedora-copr/copr/blob/7442ec4303f2e0ba2e31a6555c4da34b96de0ee1/backend/tests/test_config_reader.py#L48

We can control how many beefy machines are started from the resalloc perspective;  but we don't want to let one user to take them for the next year by submitting 10 jobs within a minute.